### ISSUE TITLE
haskell.compiler.ghc.841 is ghc 8.4.0

### ISSUE THREAD

**Issue by nmattia** [NixOS/nixpkgs#37026]
## Issue description

The attribute name is `ghc841`, I'm expecting GHC 8.4.1, not GHC 8.4.0. There might be a reason, and I might have missed it from the doc, nevertheless goes completely against the principle of least surprise.

If I want to use GHC 8.4.1, but that compiler is not packaged, I'm expecting nix to fail and tell me that the derivation does not exist yet. The current behavior is that I have to figure out why the compiler I have does not behave like the one I'm expecting.
If I want to use GHC 8.4.0, and have to use `ghc841` instead, I'll be angry at Nix the next time I update nixpkgs and end up with a compiler I don't want.

### Steps to reproduce

``` shell
$ nix-shell -p 'haskell.compiler.ghc841' --run 'ghc --version'
The Glorious Glasgow Haskell Compilation System, version 8.4.0.20180224
```

## Technical details

The commit I ran from is 91a2ad21adcccc7061c2a5f27c0f980e784b6bd9

**Comment by mpickering** [comment]
On master, it is ghc841. Previously, it was the release candidate which is versioned as you are seeing. 

It seems sensible to me to package the release candidate for ghc-8.4.1 as ghc841?

**Comment by nmattia** [comment]
I'm not sure, as outlined above I find this confusing. What are the benefits of having GHC 8.4.0, RC for GHC 8.4.1, packaged as 841 rather than 840 or 841rcX ?

**Comment by pbogdan** [comment]
In what way having 840 or 841rcX would reduce the confusion? AFAIK there's no official upstream released version of 8.4.0 and having 841rcX reporting itself as 8.4.0.YYYYMMMDD would be equally confusing?

**Comment by nmattia** [comment]
> In what way having 840 or 841rcX would reduce the confusion?

Having `840` would, IMHO, reduce the confusion because
1. It is clear that `840` is packaged. If you're looking for `840` explicitly but only see `841` you might not realize that the package you're looking for is actually there.
1. It is clear that `841` is _not_ packaged. If you're looking for `841` explicitly but only see `840` you avoid trying to figure out why you `841` compiler behaves like `840`.

Those are the two points I made above, let me know if you think they don't make sense. Regarding `841rcX`, at least the name tells you it's a release candidate and not the Real Thing.


I think there's a lot of room for personal opinions here, so I'm closing. Would be good to have a notice somewhere to warn people trying to use the un-released compiler. Not sure what a good place would be.
